Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp

```diff
@@ -124,20 +125,24 @@ int argumentsToFileDescriptor(const Vector<CString>& args, const char* name)
     return memfd;
 }
 
-static const char* applicationId(GError** error)
+static String applicationId()
 {
     GApplication* app = g_application_get_default();
-    if (!app) {
-        g_set_error_literal(error, G_IO_ERROR, G_IO_ERROR_INVALID_DATA, "GApplication is required.");
-        return nullptr;
+    if (app) {
+        const char* appID = g_application_get_application_id(app);
+        if (appID)
+            return String::fromUTF8(appID);
     }
 
-    const char* appID = g_application_get_application_id(app);
-    if (!appID) {
-        g_set_error_literal(error, G_IO_ERROR, G_IO_ERROR_INVALID_DATA, "GApplication must have a valid ID.");
-        return nullptr;
-    }
-    return appID;
+    const char* programName = g_get_prgname();
+    if (programName && g_application_id_is_valid(programName))
+        return String::fromUTF8(programName);
+
+    // There must be some id for xdg-desktop-portal to function.
+    // While xdg-desktop-portal does use this for permissions it's value
+    // doesn't matter much in our usage as we don't use any permissioned APIs.
```

Comment:
I highly doubt we can actually guarantee this comment remains true going forward, so I would reword it to instead say the best we can do is use a new app ID with new permissions rather than say that we don't do anything that requires checking permissions.

However, I worry that by using a new UUID each time we may bloat the permissions store. It's probably designed with the expectation that there will be a small number of apps. We should probably store the UUID in the data directory to avoid this. Bonus: then it will be persistent, and permissions will work properly.

Reply:
Feel fragile, because there's no way to enforce it. We won't notice if we start doing something that requires permissions, or if something we currently do that does not require a permission starts to do so in the future. The portal APIs don't have a way to assert that they'll never prompt for permission or to indicate that all permissions should be denied. But maybe we should add one....

Reply:
I'm going to merge this for now and open an issue to track this idea if that's OK?

Reply:
I suppose, though I don't see what's so bad about storing the app ID. It's not exactly a big file.

Reply:
Well, storing the identifiers opens a new can of worms: When are the files deleted? What if writing to disk fails (out of disk space, or so)? What if files disappear between runs? And so on. I would rather forget about saving the identifiers for now and see how things go. We can always make a follow-up patch later if needed.

Alternatively, it could be better to open `/proc/self/exe`, calculate the SHA256 checksum of the file, and use it as stable identifier: as long as the program on disk doesn't change, the identifier will be the same without needing to store anything else on disk... But I digress, even that I would leave for a follow up. If needed.

Reply:
This might come back to bite us in a few years when our permission stores grow huge, though. Imagine if your cookies database permanently increased in size each time you launch the browser. Using a random app ID results in the same thing, except for portal permissions.

Proposed alternative: use one hardcoded ID so that all applications get the exact same fallback ID and all permissions are shared.

Reply:
> Proposed alternative: use one hardcoded ID so that all applications get the exact same fallback ID and all permissions are shared.

If WebKitGTK happens to use a permissioned API, which is when this theoretical problem appears, then this becomes a security risk. If you grant Webcam/ScreenSharing/etc permissions to once WebContent process you grant it to all of them.

Reply:
> Alternatively, it could be better to open /proc/self/exe, calculate the SHA256 checksum of the file, and use it as stable identifier: as long as the program on disk doesn't change, the identifier will be the same without needing to store anything else on disk

This sounds like a decent idea to me.

Reply:
I suppose. A new permission store entry every time the binary changes is much less bad than a new one every time you start the program.

Reply:
https://github.com/WebKit/WebKit/pull/12761